### PR TITLE
Allow passing UI object into bundler/inline

### PR DESCRIPTION
**Pull request by lamont-granquist**
Added an argument to the gemfile method to allow overriding the
UI object.  Useful to override the default output format and
send those log events to a logfile or decorate them (timestamps,
severity, etc) before emitting them.

### CODE REVIEW

**Comment by segiddins** [comment]
I really don't think this is necessarily the right approach -- maybe a options hash that allows changing `silent` from `true` to `false`?

**Comment by lamont-granquist** [comment]
I don't want silent, though, I want my own logger.

I can certainly do an options hash instead.

**Comment by segiddins** [comment]
I'm not really sold on the benefit of you're own logger though...?

**Comment by lamont-granquist** [comment]
So I can use this in chef-client and not have messy formatting on the output for one thing.  Also, when chef-client runs as a daemon I want to capture the output and send it to the configured logfile (which is not necessarily STDOUT, particularly on Windows--runit/djb's logging philosophy is good, but its far from universal).  I don't want to suppress the output because the installer may be taking actions and not surfacing those actions to the user would be terrible experience.   I also want to be able to bump the verbosity up higher when the chef-client logger is in debug mode.

The alternative is copypasta'ing the bundler/inline code directly into chef and maintaining it there in parallel, but I'd prefer to use a public API.

**Comment by segiddins** [comment]
@indirect thoughts on taking an options hash that allows specifying a UI object?

**Comment by indirect** [comment]
seems... ok, I guess?

**Comment by lamont-granquist** [comment]
@indirect anything else you want?

File: lib/bundler/inline.rb [hunk 1]
```diff
@@ -27,7 +27,9 @@
 #
 #          puts Pod::VERSION # => "0.34.4"
 #
-def gemfile(install = false, &gemfile)
+def gemfile(install = false, options = {}, &gemfile)
+  ui = options[:ui]
```
Comment (segiddins) [inline lib/bundler/inline.rb]: should raise if there are any unknown options passed

Comment (segiddins) [inline lib/bundler/inline.rb]: `options.fetch(:ui) { Bundler::UI::Shell.new }`

**Comment by indirect** [comment]
Address @segiddins' comments, and I think this is good.

**Comment by lamont-granquist** [comment]
@segiddins @indirect here you go.

there's a bonus fix to wrap the lazy require with concurrency protection.  now that i type it up and think about it, though, i'm not positive it makes any difference since i'm not sure it makes much sense to ever have multiple threads calling into bundler?  i can back that out...

File: lib/bundler/inline.rb [hunk 2]
```diff
@@ -27,8 +27,13 @@
 #
 #          puts Pod::VERSION # => "0.34.4"
 #
-def gemfile(install = false, &gemfile)
-  require "bundler"
+def gemfile(install = false, options = {}, &gemfile)
+  Thread.exclusive { require "bundler" }
```
Comment (segiddins) [inline lib/bundler/inline.rb]: Requiring is already tread-safe, and exclusive is deprecated.

File: spec/runtime/inline_spec.rb [hunk 3]
```diff
@@ -93,4 +93,50 @@ def script(code, options = {})
...
+      end
+    RUBY
+
+    expect(out).to eq("CONFIRMED!")
+    expect(exitstatus).to be_zero if exitstatus
+  end
+
+  it "raises an exception if passed unknown arguments" do
+    script <<-RUBY, :expect_err => true
+      gemfile(true, arglebargle: true) do
```
Comment (segiddins) [inline spec/runtime/inline_spec.rb]: this fails on 1.8.7, please use hashrockets

**Comment by lamont-granquist** [comment]
@indirect travis is green

**Comment by segiddins** [comment]
@homu r+

**Pull request merged**